src/tunnel.cc

```diff
@@ -1364,7 +1364,15 @@ TunnelStateData::noteDestinationsEnd(ErrorState *selectionError)
     }
 
     // destinationsFound, but none of them worked, and we were waiting for more
-    assert(savedError);
+    debugs(17, 7, "no more destinations to try after " << n_tries << " failed attempts");
+    if (!savedError) {
+        // retryOrBail() must be preceded by saveError(), but in case we forgot:
```

Comment:
This sounds like a non-fatal error condition. Might it be worthwhile to add a debugs() to aid traceability?

Reply:
> Might it be worthwhile to add a debugs() to aid traceability?

Yes, but I would not do it _here_: We know we are in this section of the code due to debugs() immediately above this `if` statement. We know we are in this `if` statement because of fail() debugging triggered from inside of this `if` statement:

```
| 44,2| peer_select.cc(480) resolveSelected: PeerSelector1 found all 1 destinations for ...
| 17,7| FwdState.cc(650) noteDestinationsEnd: no more destinations to try after 1 failed attempts
| 17,3| FwdState.cc(461) fail: ERR_CANNOT_FORWARD "Bad Gateway"
| 17,3| FwdState.cc(185) stopAndDestroy: for all found paths have failed
```

We should add debugging to ErrorState constructor so that we do not have to wait for fail(). Done in #1768.